components/places/src/storage/bookmarks.rs

```diff
@@ -376,6 +395,259 @@ fn delete_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid) -> Result<bool> {
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct UpdatableSeparator {
+    pub location: UpdateTreeLocation,
+}
+
+impl From<UpdatableSeparator> for UpdatableItem {
+    fn from(sep: UpdatableSeparator) -> Self {
+        UpdatableItem::Separator(sep)
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct UpdatableFolder {
+    pub location: UpdateTreeLocation,
+    // There's no good reason to differentiate `null` from `""` in a folder,
+    // but for consistency we allow it.
+    pub title: UpdateTitle,
+}
+
+impl From<UpdatableFolder> for UpdatableItem {
+    fn from(folder: UpdatableFolder) -> Self {
+        UpdatableItem::Folder(folder)
+    }
+}
+
+// The type used to update the actual item.
+#[derive(Debug, Clone)]
+pub enum UpdatableItem {
+    Bookmark(UpdatableBookmark),
+    Separator(UpdatableSeparator),
+    Folder(UpdatableFolder),
+}
+
+impl UpdatableItem {
+    fn bookmark_type(&self) -> BookmarkType {
+        match self {
+            UpdatableItem::Bookmark(_) => BookmarkType::Bookmark,
+            UpdatableItem::Separator(_) => BookmarkType::Separator,
+            UpdatableItem::Folder(_) => BookmarkType::Folder,
+        }
+    }
+
+    pub fn location(&self) -> &UpdateTreeLocation {
+        match self {
+            UpdatableItem::Bookmark(b) => &b.location,
+            UpdatableItem::Separator(s) => &s.location,
+            UpdatableItem::Folder(f) => &f.location,
+        }
+    }
+}
+
+pub fn update_bookmark(db: &impl ConnExt, guid: &SyncGuid, item: &UpdatableItem) -> Result<()> {
+    let tx = db.unchecked_transaction()?;
+    let result = update_bookmark_in_tx(db, guid, item);
+    match result {
+        Ok(_) => tx.commit()?,
+        Err(_) => tx.rollback()?,
+    }
+    result
+}
+
+fn update_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid, item: &UpdatableItem) -> Result<()> {
+    let existing =
+        get_raw_bookmark(db, guid)?.ok_or_else(|| InvalidPlaceInfo::NoItem(guid.to_string()))?;
+    if existing.bookmark_type != item.bookmark_type() {
+        return Err(InvalidPlaceInfo::InvalidBookmarkType.into());
+    }
+
+    let update_old_parent_status;
+    let update_new_parent_status;
+    // to make our life easier we update every field, using existing when
+    // no value is specified.
+    let parent_id;
+    let position;
+    match item.location() {
+        UpdateTreeLocation::None => {
+            parent_id = existing.parent_id;
+            position = existing.position;
+            update_old_parent_status = false;
+            update_new_parent_status = false;
+        }
+        UpdateTreeLocation::Position(pos) => {
+            parent_id = existing.parent_id;
+            update_old_parent_status = true;
+            update_new_parent_status = false;
+            // Not clear that InvalidParent is the correct error here - probably
+            // should be a "stuff is corrupt" error? Or maybe we should fix it
+            // here by reparenting to unfiled?
```

Comment:
I don't think we should reparent automatically, but I think a corruption error makes sense! Let's throw so the consumer can catch it, and eventually run our as-yet-unimplemented maintenance.